### ISSUE TITLE
Leaky Bucket Limiter

### ISSUE THREAD

**Issue by ideaoforder** [mperham/sidekiq#4414]
When limiting against APIs that use a Leaky Bucket rate limit, it'd be helpful to limit requests in Sidekiq using a Leaky Bucket Limiter.

References:
https://help.shopify.com/en/api/reference/rest-admin-api-rate-limits#general-api-rate-limits
https://en.wikipedia.org/wiki/Leaky_bucket

**Comment by jeffblake** [comment]
+1 for leaky bucket

Another piece with Shopify, is Plus companies have twice the bucket size (40) vs regular customers (20). I don't see see how another redis or database call could be avoided to get that number, but perhaps the wizard Mike has an idea. 

Maybe the bucket size could be set once elsewhere (upon account creation), and if it isn't specified directly in the Limiter, it defaults to that value.

E.g.

Customer signs up
`Sidekiq::Limiter.leaky_bucket("shopify-#{shop_domain}", size: 40, duration: 2.minutes)`

Called later in a worker
// Uses 40, per 2 min
`Sidekiq::Limiter.leaky_bucket("shopify-#{shop_domain}", default: 20 per 2.minutes)`

Or 
// explicitly uses 20 per 2 min
`Sidekiq::Limiter.leaky_bucket("shopify-#{shop_domain}", size: 20, duration: 2.minutes)`

**Comment by mperham** [comment]
I'm looking into this now and it's not clear how to implement every detail of the leaky bucket. If the policy is "50 calls every 10 seconds", should I allow a burst of N (50? 20?) calls to enter the bucket quickly? Should callers be told when the next slot will be available so they can sleep until then or just raise OverLimit?

I really need detailed use cases and expected behavior to nail down the edge cases. Knowing Shopify's exact limiter logic would help but that public doc doesn't give much detail.

**Comment by mgomes** [comment]
@mperham that documentation seems to have deteriorated over time. It used to give an example of the request rate and also the bucket limit for each.

So hypothetically (this used to be Shopify's limit and it might still be), given a 2 reqs/sec limit with a bucket limit of 30 requests, I can burst up to 30 and each second thereafter another 2 jobs can be run.

Therefore we wouldn't other jobs to immediately raise `OverLimit` but instead wait around for some timeout amount. I think the limiter would need 4 arguments: `bucket size`, `interval`, `rate`, and `wait_timeout`. So similar to the current bucket limiter but with a max bucket size.

A Shopify use case would be updating say 20 products. Rather than doing it over 10 seconds, an app could process all 20 right away. But if I had to process 300, I could burst 30 right away and each second thereafter I could process an additional 2.

**Comment by edavis10** [comment]
@mgomes is right except the bucket limits are 40 requests (80 for enterprise accounts).

If you have 40 requests to make, they can happen immediately with no delay.

If you have 41 requests though, the first 40 would process immediately while the 41st would stall out with HTTP 429 code (Rate Limit Exceeded). Wait half a second and the API bucket would fill up by one allowing the 41st in.

These are also per store-account, so example1.com can have 40 requests running for it at the same time as 40 for example2.com and 40 for example3.com. (e.g. running against stores in parallel is much faster than parallel against each store in sequence). This is the big opportunity for a major speedup.

![image(1)](https://user-images.githubusercontent.com/1621/98055539-8e341080-1df2-11eb-9070-4d3bd30a8739.png)

Here's a screenshot of this from my monitoring system. You can see there are many successful requests (beige counter) going and any above that hit the limit (blue counter). Each window is one minute so the 2 requests/sec limiter becomes the 120 you see. The rest of the requests hit the limiter and in my code just `sleep` and retry. The 120 requests/minute continue until they clear.

In Sidekiq all of the jobs just run as fast as possible in my apps and the `sleep` code causes the job to be put on hold while the API limit refills. This occupies a sidekiq thread while it waits.

It would be much better to be able to tell Sidekiq how fast to process the jobs and let all of the API limited threads do other work (e.g. other store's jobs).

**Comment by Alagaesia93** [comment]
Jumping in a very old issue because I'm exploring how to do proper rate limiting for Shopify's GraphQL Admin API. Documentation link => https://shopify.dev/api/usage/rate-limits#graphql-admin-api-rate-limits

I read about the leaky bucket algorithm and how it's implemented in [sidekiq ent](https://github.com/mperham/sidekiq/wiki/Ent-Rate-Limiting), but it's possible only to have a number of API calls per unit of time. Shopify's GQL works totally different, it's point based, and every API request returns the remaining points you have, and they fill up again every N seconds.

How would you approach this? Happy to open a PR with some general solution, if you help me find one @mperham! 

Thanks a lot in advance!

**Comment by kirillplatonov** [comment]
@Alagaesia93 opened a new issue for that https://github.com/mperham/sidekiq/issues/5757
